R/modules.R

```diff
@@ -353,6 +241,157 @@ module <- function(label = "module",
   )
 }
 
+#' @rdname teal_modules
+#' @export
+#'
+modules <- function(..., label = "root") {
+  checkmate::assert_string(label)
+  submodules <- list(...)
+  if (any(vapply(submodules, is.character, FUN.VALUE = logical(1)))) {
+    stop(
+      "The only character argument to modules() must be 'label' and it must be named, ",
+      "change modules('lab', ...) to modules(label = 'lab', ...)"
+    )
+  }
+
+  checkmate::assert_list(submodules, min.len = 1, any.missing = FALSE, types = c("teal_module", "teal_modules"))
+  # name them so we can more easily access the children
+  # beware however that the label of the submodules should not be changed as it must be kept synced
+  labels <- vapply(submodules, function(submodule) submodule$label, character(1))
+  names(submodules) <- make.unique(gsub("[^[:alnum:]]+", "_", labels), sep = "_")
+  structure(
+    list(
+      label = label,
+      children = submodules
+    ),
+    class = "teal_modules"
+  )
+}
+
+# printing methods ----
+
+#' Converts `teal_module` to a string
+#'
+#' @inheritParams format.teal_modules
+#' @param x (`teal_module`)
+#' @rdname teal_modules
+#' @export
+format.teal_module <- function(x, indent = 0, ...) { # nolint
+  paste0(paste(rep(" ", indent), collapse = ""), "+ ", x$label, "\n", collapse = "")
+}
+
+
+#' Prints `teal_module`
+#' @param x (`teal_module`)
```

Comment:
@chlebowa after 

```r
@param x (`teal_module`)
```
there is another

```r
@param x (`teal_modules`)
```

in `print.teal_modules` and this gets overvritten by the last appearance of `@param x`

Should you document once? This also covers the documentation of `format`, since all 4 functions (2 formats and 2 prints) are on the same page.

Reply:
af6c569a2cfd978eacaf4a9ee57f7d4bc08b3af9

Reply:
I collated them under one `x`.